unstructured_inference/models/chipper.py

```diff
@@ -1,173 +1,502 @@
+                        y_offset,
+                        ratio,
+                        image.size,
+                    )
+
+                    element.bbox = Rectangle(*bbox_coords)
 
-                    elements.append(LocationlessLayoutElement(type=stype[3:-1], text=string))
+                    self.update_parent_bbox(element)
 
                 start = -1
-                last_special_token = output_ids[i]
             else:
                 if start == -1:
                     start = i
 
                 end = i
 
         # If exited before eos is achieved
-        if start != -1 and start < end:
+        if start != -1 and start < end and len(parents) > 0:
             slicing_end = end + 1
             string = self.tokenizer.decode(output_ids[start:slicing_end])
 
-            stype = self.tokenizer.decode(last_special_token)
+            element = parents.pop(-1)
+            element.text = string
 
-            elements.append(LocationlessLayoutElement(type=stype[3:-1], text=string))
+            bbox_coords = self.get_bounding_box(
+                decoder_cross_attentions=decoder_cross_attentions,
+                tkn_indexes=list(range(start - 1, end)),
+                final_w=self.processor.image_processor.size["width"],
+                final_h=self.processor.image_processor.size["height"],
+                heatmap_w=self.heatmap_w,
+                heatmap_h=self.heatmap_h,
+            )
+
+            bbox_coords = self.adjust_bbox(
+                bbox_coords,
+                x_offset,
+                y_offset,
+                ratio,
+                image.size,
+            )
+
+            element.bbox = Rectangle(*bbox_coords)
+
+            self.update_parent_bbox(element)
 
         return elements
+
+    def adjust_bbox(self, bbox, x_offset, y_offset, ratio, target_size):
+        """Translate bbox by (x_offset, y_offset) and shrink by ratio."""
+        return [
+            max((bbox[0] - x_offset) / ratio, 0),
+            max((bbox[1] - y_offset) / ratio, 0),
+            min((bbox[2] - x_offset) / ratio, target_size[0]),
+            min((bbox[3] - y_offset) / ratio, target_size[1]),
+        ]
+
+    def get_bounding_box(
+        self,
+        decoder_cross_attentions: Sequence[Sequence[torch.Tensor]],
+        tkn_indexes: List[int],
+        final_h: int = 1280,
+        final_w: int = 960,
+        heatmap_h: int = 40,
+        heatmap_w: int = 30,
+        discard_ratio: float = 0.99,
+    ) -> List[int]:
+        """
+        decoder_cross_attention: tuple(tuple(torch.FloatTensor))
+        Tuple (one element for each generated token) of tuples (one element for
+        each layer of the decoder) of  `torch.FloatTensor` of shape
+        `(batch_size, num_heads, generated_length, sequence_length)`
+        """
+        agg_heatmap = np.zeros([final_h, final_w], dtype=np.uint8)
+
+        for tidx in tkn_indexes:
+            hmaps = torch.stack(list(decoder_cross_attentions[tidx]), dim=0)
```

Comment:
why do we need to cast into `list` first? Can we require this to just be higher dim tensor so that we don't need to do type casting?